### DISCUSSION TITLE
How to use OptiStack.solver()?

### DISCUSSION THREAD

**Discussion by kimanton** [casadi/casadi#3571] [Q&A]
I am trying to write an MPC for UR10, and I wanted to see if I can do it with casadi

As in the title, I don't understand how to properly use solver()  
The documentation says it needs two dictionaries, one with options for nlpsol solver, and another for its plugin
```
solver(self, str solver, dict plugin_options, dict solver_options)  
   Set a solver.  
   Parameters:  
      solver: any of the nlpsol plugins can be used here In practice, not all nlpsol plugins may be supported yet  
      options: passed on to nlpsol plugin No stability can be guaranteed about this part of the API  
      options: to be passed to nlpsol solver No stability can be guaranteed about this part of the API
```  
However when I try to provide two dictionaries I always get error `"Unknown option: sqpmethod"`
For example, this is how I did it initially
```
qp_opts = {"osqp": {'max_iter':30}}
p_opts = {"qpsol": 'osqp', "max_iter": 10, 'qpsol_options':qp_opts}
s_opts = {"error_on_fail": False}
opti.solver("sqpmethod", p_opts, s_opts)
```
And this is the error message
```
RuntimeError: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:157:
.../casadi/core/options.cpp:244: Unknown option: sqpmethod
```

I then read [this](https://github.com/casadi/casadi/discussions/3541) discussion. The author there used only one dictionary and it seems to work. I modified my code to this
```
qp_opts = {"osqp": {'max_iter':3}}
p_opts = {"qpsol": 'osqp', "max_iter": 1,"error_on_fail": False, 'qpsol_options':qp_opts}
opti.solver("sqpmethod", p_opts)
```
And received the following error
```
RuntimeError: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:157:
Error in Function::call for 'solver' [Sqpmethod] at .../casadi/core/function.cpp:1401:
Error in Function::call for 'solver' [Sqpmethod] at .../casadi/core/function.cpp:330:
Error in Function::operator() for 'qpsol' [OsqpInterface] at .../casadi/core/function.cpp:1482:
.../casadi/core/conic.cpp:537: conic process failed. Set 'error_on_fail' option to false to ignore this error.
```

I then found [this](https://github.com/casadi/casadi/discussions/3221) discussion. jgillis suggested that maybe problem is getting non convex and they should try to use regularization. So I did
```
qp_opts = {"osqp": {'max_iter':3}}
p_opts = {"qpsol": 'osqp', "max_iter": 1, 'convexify_strategy':'regularize',"error_on_fail": False, 'qpsol_options':qp_opts}
opti.solver("sqpmethod", p_opts)
```
But still get the same error as before 
```
.../casadi/core/conic.cpp:537: conic process failed. Set 'error_on_fail' option to false to ignore this error.
```

What am I doing wrong?

**Comment by brunoml5** [comment]
Hi!

For sqpmethod, you need to give options for the NLP solver and the underlying QP solver. The correct syntax is the following:
```
qp_opts = {"error_on_fail": False}
p_opts = {"qpsol": 'osqp', "max_iter": 1, 'convexify_strategy':'regularize',
          'qpsol_options':qp_opts}

opti.solver("sqpmethod", p_opts)
```
Or you can do:
```
p_opts = {"qpsol": 'osqp', "max_iter": 1, 'convexify_strategy':'regularize',
          'qpsol_options.error_on_fail': False}

opti.solver("sqpmethod", p_opts)
```
Indeed, the documentation is not very clear about this.

**Reply by kimanton** [reply]
Thank you, I think it helped
However now I get another error message
```
RuntimeError: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:157:
.../casadi/core/optistack_internal.cpp:996: Assertion "return_success(accept_limit)" failed:
Solver failed. You may use opti.debug.value to investigate the latest values of variables. return_status is 'Maximum_Iterations_Exceeded'
```
At first it seems that it didn't converge and exceeded maximum number of iterations, but I tried increasing it iterations to 20. After 134s it still gave me the same error.

Does it suggest that my problem formulation is poor or I just didn't choose right solver or solver options?

I also wanted to check how it would do with qpoases
I can't get it to stop, it continues execution indefinitely although I added multiple limiting options `"CPUtime": 1, 'numRefinementSteps':10, 'numRegularisationSteps':10, 'terminationTolerance':0.1` 
This again makes me think that maybe I am doing something wrong

**Comment by brunoml5** [comment]
> 
> Does it suggest that my problem formulation is poor or I just didn't choose right solver or solver options?
> 
Probably both.

The developers said somewhere that the `sqpmethod` is a textbook method, it is not very robust or stable. Be careful with the initial guess. Maybe solve your problem first without objective function and then use the result as a initial guess for the `sqpmethod`. Then, the solver will start from a feasible solution.

If you really wanna use a SQP method, try a professional solver such as WORHP or SNOPT. Or maybe try `sleqp` which is available in version 3.6.4 and is not on the documentation. Not available yet in conda-forge though (for some reason), you need to install through pip.
